rfcs/0001-messages.md

```diff
@@ -0,0 +1,260 @@
+    Commitments(MsgCommitments),
+    SigningPackage(MsgSigningPackage),
+    SignatureShare(MsgSignatureShare),
+    FinalSignature(MsgFinalSignature),
+}
+```
+
+All the messages and new types will be defined in a new file `src/frost/messages.rs`
+
+## Reference-level explanation
+
+Here we explore in detail the header types and all the message payloads. 
+
+### Header
+
+Fields of the header define new types. Proposed implementation for them is as follows:
+
+```rust
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    DealerBroadcast,
+    Commitments,
+    SigningPackage,
+    SignatureShare,
+    FinalSignature,
+}
+
+struct MsgVersion(u8);
+
+struct Participant(u8);
+```
+
+### Payloads
+
+Each payload defines a new message:
+
+```rust
+// Dealer must send this message with initial data to each participant involved.
+// With this, the participant should be able to build a `SharePackage` and use
+//  the `sign()` function.
+// `group_public` is random data at this stage and `public_key` can be calculated
+// from `secret_key`.
+struct MsgDealerBroadcast {
+    // The secret key as a frost::Scalar.
+    secret_key: frost::Scalar,
+    // Set of commitments as jubjub::ExtendedPoint using frost::Commitment wrapper.
+    commitments: Vec<frost::Commitment>,
+    // The generated public key for the group.
+    group_public: VerificationKey<SpendAuth>,
+}
+
+// Each signer participant send to the aggregator the 2 points
+//  needed for commitment building.
+struct MsgCommitments {
+    // The hiding Point.
+    hiding: jubjub::ExtendedPoint,
+    // The binding Point.
+    binding: jubjub::ExtendedPoint,
+}
+
+// The aggergator decide what message is going to be signed and
+// send it to each participant with all the commitments collected.
+struct MsgSigningPackage {
+    // Consist of a [u8] and vector of commitments.
+    // Commitments are a signer id and a hiding and binding point.
+    signing_package: frost::SigningPackage,
+}
+
+// Each signer send the signatures to the agregator who is going to collect them 
+// and generate a final spend signature.
+struct MsgSignatureShare {
+    // The signature to be shared as a Scalar
+    signature: frost::Scalar,
+}
+
+// The final signature is broadcasted by the aggegator 
+// to any participant.
+struct MsgFinalSignature {
+    final_signature: Signature<SpendAuth>,
+}
+```
+
+## Validation
+
+Validation is implemented to each new data type as needed. This will ensure the creation of valid messages before they are sent and right after they are received. We create a trait for this as follows:
+
+```rust
+pub trait Validate {
+    fn validate(&self) -> &Self;
+}
+```
+
+And we implement where needed. For example, in the header sender and receiver can't be the same:
+
+```rust
+impl Validate for Header {
+    fn validate(&self) -> &Self {
+        if self.sender.0 == self.receiver.0 {
+            panic!("sender and receiver are the same");
+        }
+        self
+    }
+}
+```
```

Comment:
Panicking creates a trivial denial of service attack on the receiver: just send an invalid message.

Instead, return a `Result`, and define an `Error` type. Then you can panic in the sender using `expect`.

In the receiver you can just ignore the message. Since it's a threshold signing scheme, you can ignore some messages from some participants, and still succeed. But the messages from each sender must be optional, and the caller must put a timeout on receiving each message. (Otherwise the caller could hang waiting for a participant that has crashed or has an invalid message version.)

Reply:
The simplest error type is just an enum with the different kinds of errors you expect. See Zebra for some good examples.

Reply:
Error type and such is done. We are not going to implement any timer at this stage.

Reply:
Ok, so you'll just hang if one of the participants in the round goes down?
That should be documented.

Reply:
https://github.com/ZcashFoundation/redjubjub/pull/85/commits/104d0f59223e9580fa51141307a4e397eedadf36